docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,81 @@
+---
+title: Global / Federated Rules API
+type: proposal
+menu: proposals
+status: accepted
+owner: s-urbaniak,bwplotka
+---
+
+### Related Tickets
+
+https://github.com/thanos-io/thanos/issues/1375
+
+## Summary
+
+Thanos allows a global query view for the Prometheus series. This means discovering, connecting various (often remote) "leafs" components and aggregating series data from them.
+
+Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.
+
+## Motivation
+
+Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it ThanosRuler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* It can be tied to Thanos.
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy RulesAPI and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+We propose to add following proto service and propagate all using existing components that has rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
```

Comment:
Why stream? Wouldn't it me simpler to do an array?

Reply:
The intent was here to be in sync with Store API and potentially be prepared for future enhancement to be potentially stream based.

I do agree though, that in the case of rules this could potentially be a premature fore-sight.

Reply:
I see the point of @povilasv here, but I fully agree with @s-urbaniak here.

My arguments for streaming: 
* Readability and consistency wins here as we already have tools for streaming dedup, partial response, timeouts and iterators so we can reuse more and all the patterns.
* It's will use bit less resources (e.g const merge time), but it's true as @povilasv said that we are not talking about high load here.

Reply:
ok, resolving the conversation as it seems we have reached consensus :-)